qcodes/utils/dataset/doNd.py

```diff
@@ -483,6 +483,180 @@ def do2d(
     return _handle_plotting(dataset, do_plot, interrupted())
 
 
+def dond(
+    *params: Union[_BaseParameter, int, float, ParamMeasT],
+    write_period: Optional[float] = None,
+    measurement_name: str = "",
+    exp: Optional[Experiment] = None,
+    do_plot: Optional[bool] = None,
+    show_progress: Optional[None] = None,
+    use_threads: bool = False,
+    additional_setpoints: Sequence[ParamMeasT] = tuple()
+        ) -> AxesTupleListWithDataSet:
+    """
+    Perform n-dimentional scan from slowest (first) to the fastest (last), to
+    measure m measurement parameters. Supplied params are parsed to params_set,
+    params_dicts and params_meas inside the function.
+
+    Args:
+        *params:
+            param_set_1, start_1, stop_1, num_points_1, delay_1, ...,
+            param_set_n, start_n, stop_n, num_points_n, delay_n,
+            param_meas_1, param_meas_2, ..., param_meas_m
+        write_period: The time after which the data is actually written to the
+            database.
+        measurement_name: Name of the measurement. This will be passed down to
+            the dataset produced by the measurement. If not given, a default
+            value of 'results' is used for the dataset.
+        exp: The experiment to use for this measurement.
+        do_plot: should png and pdf versions of the images be saved and plots
+            are shown after the run. If None the setting will be read from
+            ``qcodesrc.json``
+        show_progress: should a progress bar be displayed during the
+            measurement. If None the setting will be read from ``qcodesrc.json`
+        use_threads: If True, measurements from each instrument will be done on
+            separate threads. If you are measuring from several instruments
+            this may give a significant speedup.
+        additional_setpoints: A list of setpoint parameters to be registered in
+            the measurement but not scanned.
+    """
+    if do_plot is None:
+        do_plot = config.dataset.dond_plot
+    if show_progress is None:
+        show_progress = config.dataset.dond_show_progress
+
+    meas = Measurement(name=measurement_name, exp=exp)
+
+    def _make_nested_setpoints(
+        params: List[_BaseParameter],
+        param_dicts: Dict[_BaseParameter, Dict[str, Union[float, int]]]
+    ) -> Union[np.ndarray, List[Tuple[()]]]:
```

Comment:
No that is not allowed. We are using disallow_any_generics = True so all Any needs to be written out explicitly. 
This should probably be `List[Tuple[Any, ...]]`

Reply:
For do0d, the nested setpoint should be an empty tuple. I did look at the documentation and it is stated like this:
"The type of the empty tuple can be written as Tuple[()]."
https://docs.python.org/3/library/typing.html#typing.Tuple

Reply:
I couldn't get this last suggestion to work. we need an object to be able to iterate over one time for dond_0d. So somethig like this also works `[[]]` or `[np.array([])]` or `(np. ndarray([]))`, etc.
I took the if statement out of this private function, though and setup the similar logic, when I wanted to get the return of this function. Like this
```
if sweep_instances:
        nested_setpoints = _make_nested_setpoints(sweep_instances)
    else:
        nested_setpoints = [()]  # For zero-dimensional grid (do0d)
```
I don't know if you have any better way to dela with this.

Reply:
the loop inside ``with meas.run()`` that calls ``datasaver.add_results`` is implemented in such a way that in order for it to work for 0D cases, nested_setpoints must have a single empty-tuple (or emplty-sequence) value -- it semantically means "do a single 'get' of the measured parameter(s), but there are no specific setpoint values for this 'get'". The reason _make_nested_setpoints can't give that for us automatically is because ``np.vstack`` does not accept an empty list as an input (that's what happens when the list of the given sweeps is empty).

My opinion is that it's OK to let ``_make_nested_setpoints`` return ``np.array([[]])`` (empty list in a list) in case ``flat_setpoint_grids`` list inside the functions is empty. This will remove the need for the ``if sweep_instances`` clause altogether, and keeps the code readable. see here:
![image](https://user-images.githubusercontent.com/15662810/120801587-a06d5d00-c541-11eb-81fb-379b493a6130.png)

i tested this locally by calling dond with an arrayparameter, and it works and propagates the shapes correctly.

would you agree @FarBo ?

Reply:
Thanks for the explanation @astafan8 
So it explains why I couldn't get it to work with an empty numpy array.

I see you have nicely used numpy array to hide a list of list :). If it works, then makes the the nested_setpoints more consistant and I like it :). Perhaps I only change the if statement to not check the flat_setpoint_grid, but rather setpoint_values as it is also an empty list if there is no set parameters in dond.

I would try your suggestion locally and run my tests on it to see what I get and adapt to that if no problem found :)